src/librustc_parse/parser/item.rs

```diff
@@ -1740,6 +1742,26 @@ impl<'a> Parser<'a> {
         ).emit();
     }
 
+    /// Checks if current token is keyword that is currently being parsed and reports an error if it is.
+    fn report_nested_keyword(&mut self, kw: Symbol) -> PResult<'a, ()> {
+        if self.token.is_keyword(kw) {
```

Comment:
Instead of taking in `kw: Symbol`, I think it would be better to check whether the current token is one of `kw::Enum`, `kw::Struct`, or `kw::Union`, use `let item = self.parse_item()?;` to parse the full nested `enum`, `struct`, or `union`, and then finally use `item.span`.

You can use `self.token.clone()` + `pprust::token_to_string` as a substitute for `kw` in the `format!(...)`s below.

Reply:
The problem with doing `let item = self.parse_item()?;` is that we can have a case:
```rust
enum Foo {
    enum, // someone forgot to capitalize 
    bar,
}
```
and this could potentially produce errors which are not really relevant to the code.

Reply:
Parser recovery is all about making educated guesses as what to what the user meant. It's unlikely that the user really wanted to name a variant `enum` -- if you truly want to catch that case, you can look-ahead 1 token for `token::Comma` and avoid doing `self.parse_item()?;` but I think it's not worth the extra complexity.

Meanwhile, doing `self.parse_item()?` makes for better recovery in the common case because you turn fatal parser errors into recoverable ones so that compilation can continue to e.g. type checking and so on.